File: node-js-lambda/lambda/src/handlers/storeUser.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';

const dynamo = DynamoDBDocumentClient.from(new DynamoDBClient({}));

const TABLE_NAME = process.env.USERS_TABLE_NAME!;

interface StoreUserBody {
  name: string;
  email: string;
}

export interface User {
  userId: string;
  name: string;
  email: string;
  createdAt: string;
}

function generateUserId(): string {
  return `usr_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
}

export async function storeUser(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
  if (!event.body) {
    return json(400, { error: 'Request body is required' });
  }

  let body: StoreUserBody;
  try {
    body = JSON.parse(event.body);
  } catch {
    return json(400, { error: 'Invalid JSON body' });
  }

  const { name, email } = body;
  if (!name || !email) {
    return json(400, { error: 'name and email are required' });
  }

  const userId = generateUserId();
  const user: User = {
    userId,
    name,
    email,
    createdAt: new Date().toISOString(),
  };

  console.log(`Storing user with id: ${userId}.`);

  try {
    await dynamo.send(new PutCommand({ TableName: TABLE_NAME, Item: user }));
  } catch (err) {
    const error = err as Error;
    console.error('Failed to store user', { error: error.message, stack: error.stack });
    return json(500, { error: 'Internal server error' });
  }

  console.log(`User stored successfully with id: ${userId}`);
  return json(201, user);
}

function json(statusCode: number, body: unknown): APIGatewayProxyResult {
  return {
    statusCode,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  };
}

File: python-lambda/lambda/handlers/store_user.py
import json
import logging
import os
import random
import string
import time
from datetime import datetime, timezone

import boto3

logger = logging.getLogger(__name__)

dynamo = boto3.resource('dynamodb')
TABLE_NAME = os.environ['USERS_TABLE_NAME']


def _generate_user_id() -> str:
    suffix = ''.join(random.choices(string.ascii_lowercase + string.digits, k=7))
    return f"usr_{int(time.time() * 1000)}_{suffix}"


def store_user(event: dict) -> dict:
    if not event.get('body'):
        return _json(400, {'error': 'Request body is required'})

    try:
        body = json.loads(event['body'])
    except (json.JSONDecodeError, TypeError):
        return _json(400, {'error': 'Invalid JSON body'})

    name = body.get('name')
    email = body.get('email')
    if not name or not email:
        return _json(400, {'error': 'name and email are required'})

    user_id = _generate_user_id()
    user = {
        'userId': user_id,
        'name': name,
        'email': email,
        'createdAt': datetime.now(timezone.utc).isoformat(),
    }

    logger.info(f"Storing user with id: {user_id}.")

    dynamo.Table(TABLE_NAME).put_item(Item=user)

    logger.info(f"User stored successfully with id: {user_id}")
    return _json(201, user)


def _json(status_code: int, body: object) -> dict:
    return {
        'statusCode': status_code,
        'headers': {'Content-Type': 'application/json'},
        'body': json.dumps(body),
    }
